### PR TITLE
libp2p adjustments

### PR DESCRIPTION
**Pull request by mpetrunic**
I've exposed libp2p as public prop and it can be passed override creating new.
this way we can reuse existing infra for opts overriding

### CODE REVIEW

**Comment by mpetrunic** [comment]
@weemeetagain

**Review by twoeths — APPROVED** [review]

File: packages/lodestar/src/network/network.ts [hunk 1]
```diff
@@ -8,59 +8,48 @@ import promisify from "promisify-es6";
...
+import {initializePeerInfo, initPeerId} from "./util";
+import {NodejsNode} from "./nodejs";
+import deepmerge from "deepmerge";
 
-interface ILibp2pModules {
+
+interface INetworkModules {
   config: IBeaconConfig;
-  libp2p: LibP2p;
+  libp2p?: LibP2p;
```
Comment (ec2) [inline packages/lodestar/src/network/network.ts:24]: woah. is this our first use of optionals???? :O

Comment (mpetrunic) [reply packages/lodestar/src/network/network.ts:24]: thats not new chaining optionals, this is regular optional :)

File: packages/lodestar/src/network/util.ts [hunk 2]
```diff
@@ -8,7 +8,9 @@ import PeerInfo from "peer-info";
 import promisify from "promisify-es6";
 import {Shard} from "@chainsafe/eth2.0-types";
 
-import {RequestId, SHARD_SUBNET_COUNT, SHARD_ATTESTATION_TOPIC, BLOCK_TOPIC, ATTESTATION_TOPIC} from "../constants";
+import {ATTESTATION_TOPIC, BLOCK_TOPIC, RequestId, SHARD_ATTESTATION_TOPIC, SHARD_SUBNET_COUNT} from "../constants";
+import fs from "fs";
+import {loadPeerId} from "./nodejs";
```
Comment (wemeetagain) [inline packages/lodestar/src/network/util.ts:13]: I was trying to keep the nodejs specific code in ./nodejs
Now this is importing the fs module into a non-nodejs-specific module

File: packages/lodestar/src/network/network.ts (hunk 1 above)
Comment (ec2) [reply packages/lodestar/src/network/network.ts:24]: Oh cool. Never used it before. Didn't notice em before in our code either 😅

**Comment by ec2** [comment]
@mpetrunic Could you explain the rationale behind this a bit more? I'm not entirely sure why we would want to couple libp2p into node. Seems like the change here is instead of creating libp2p in `dev.ts`, we create it inside of the beacon node. Not sure why this is better...

**Review by wemeetagain — CHANGES_REQUESTED** [review]
IMO this not actually an improvement.
I think we should just keep the code as is, having the libp2p object passed in as a dependency.

It's nice to have the libp2p initialization logic decoupled from the network module. It also matches the logger and eth1 as external resources to be configured independently of the BeaconNode. This PR does a similar thing, by allowing libp2p to be passed in, but now couples the initialization to the network module. My leaning would be to add these sorts convenience functions around the outside (of the BeaconNode initialization) rather than pollute the core modules.

**Comment by mpetrunic** [comment]
The rationale is that we are initializing libp2p in 2 or three places along with peerId initialization and  thats just code duplication. We also breka our optionts arhitecture and it's quite out of place.

I agree that peerId from file should not be there. We could move that to cli process logic.

Why would be wan't to keep libp2p outside of networking module?

File: packages/lodestar/src/network/util.ts (hunk 2 above)
Comment (mpetrunic) [reply packages/lodestar/src/network/util.ts:13]: yeah agree, it should be easy to move into cli process. Check if it's a file and load as string

**Comment by mpetrunic** [comment]
One other improvment would be to use some kind of creator parrent. Like:
```
interface Libp2pCreator {
   create(): Promise<Libp2p>
}
```
```
class NodeJsLibp2pCreator implements Libp2pCreator
class WebLibp2pCreator implements Libp2pCreator
```
and one of them would be passed to network as module.

I can update this PR with this changes

**Comment by ec2** [comment]
I thought we only init libp2p once in the `dev.ts`. I couldn't find anywhere else we initialize libp2p

**Comment by ec2** [comment]
Also, in my mind i have a very very hypothetical situations. But say for example i wanna spin up a bunch of beacon nodes sharing one libp2p connection, this could become a lot harder. Just a thought if someone wants to do some horizontal scaling or something. VERY hypothetical situation.

**Comment by wemeetagain** [comment]
We can deduplicate the code in different ways, I like @tuyennhv 's idea of collapsing the libp2p initialization into a utility function.
We can abstract this into a similar thing, and then this PR becomes a sort of convenience (to initialize a libp2p when one is not passed in).

The vision for the BeaconNode that I have is that it is a "low level" interface that doesn't know about the external world or external resources in itself. It only knows about how to drive a beacon chain, and use well-defined interfaces with the outside world to do so. It is given references to those things: the network, the eth1 notifier, etc, because outside parties are better suited at configuring those resources, knowing more about the environment the BeaconNode finds itself in.

In this way of doing things, there would be convenience functions that wrap around _outside_ the BeaconNode that provide the convenience of using more well-known configurations for external resources. For example, an end user may run a `createNodejsBeaconNode(opts)` to initialize a BeaconNode with some standard configuration, and that implementation would pick the right resources for the BeaconNode. This is in keeping with the style of other internal modules (eg: `db` being passed in to `chain`) where dependencies are passed in and used according to their public interfaces, but are independently configured.

**Comment by mpetrunic** [comment]
Yeah but he should configure it using opts and not by need to pull out deep internal code (libp2p, peerId). If he want's to switch to web beacon node, he ain't suppose to duplicate code for creating peerId, peerInfo and all other stuff that can already be changed by opts. Creator pattern is much more suited as it allows to reuse code for common scenarios.

The same code is in dev and regular beacon node command and it will be in e2e test and probably more stuff. as we go along

is Libp2p separate module or is it part of networking. It cannot stay semi module with 20 lines of code needed to initialize it

**Comment by wemeetagain** [comment]
If they want to switch to web beacon node, they will need a libp2p object that is configured much differently than a nodejs-configured one. So they will need to duplicate code in so far as they will need to create a web-suitable libp2p object. The BeaconNode should know nothing of these details. Ofc in most scenarios, a high level function, eg: `createWebBeaconNode` would handle a default configuration.

**Comment by mpetrunic** [comment]
They will just pass different opts that configures libp2p like they wan't. Creator pattern can enable them to change core stuff, most of the stuff is the same anyways. 

If someone does need some extremely different cases, they can still pass custom Libp2p object.

It makes no sense to break our configuration pattern here and copy paste same code in 3 different places

**Pull request closed without merging**